### ISSUE TITLE
AuthState observable issue when a new user registers

### ISSUE THREAD

**Issue by uncvrd** [angular/angularfire#2019]
### Version info

**Angular:**
7.2.2

**Firebase:**
5.0.0-beta23

**AngularFire:**
5.1.1

**Other (e.g. Ionic/Cordova, Node, browser, operating system):**
Ionic:

```
   ionic (Ionic CLI)             : 4.10.3 (/usr/local/lib/node_modules/ionic)
   Ionic Framework               : @ionic/angular 4.0.2
   @angular-devkit/build-angular : 0.12.2
   @angular-devkit/schematics    : 7.2.2
   @angular/cli                  : 7.2.2
   @ionic/angular-toolkit        : 1.2.2

Cordova:

   cordova (Cordova CLI) : 8.0.0
   Cordova Platforms     : android 6.4.0, ios 4.5.5
   Cordova Plugins       : cordova-plugin-ionic-keyboard 2.1.3, cordova-plugin-ionic-webview 2.3.1, (and 9 other plugins)

System:

   ios-deploy : 1.9.2
   NodeJS     : v10.15.0 (/usr/local/bin/node)
   npm        : 6.4.1
   OS         : macOS Mojave
   Xcode      : Xcode 10.1 Build version 10B61
```

### How to reproduce these conditions

Recently, I switched from Firestore to a custom MySQL database to hold data, but decided to continue using Firebase Auth within the app. I use the `AuthState` observable to determine if a user is logged in, within my `AuthService` constructor like so:

```
this.user$ = this.afAuth.authState.pipe(
      switchMap(
        (user) => {
          return user ? db.query$(`user/find?userUID=${user.uid}`) : of(null)
        } 
      )
    );
```

I access the `user$` variable between several pages to determine when a user is logged in. For example, on my home page, I display a certain `div` when logged in like so:

**Home.page.ts:**

```
ngOnInit(){
    //where auth is my instance to my AuthService
    this.user$ = this.auth.user$;

    //I also subscribe to the auth.user$ below to pull in data for something else like so
    this.auth.user$.subscribe((user) => {
        //do stuff with user
    }
}
```

I then use the `this.user$` variable defined in my `ngOnInit` to asynchronously display a container

**Home.page.html**

```
<ng-container *ngIf="user$ | async as user; else unAuth">
//content
</ng-container>
```

This works flawlessly when using the log in and log out features of AngularFire, however I have run in to an issue when using this method for **registering** a user. For some reason, on registration, the observable on the home page does not pick up the state change. Once I refresh the page, everything works as expected, but of course that's a less than desirable action on a SPA.

To make sure my service was detecting the state change on registration, I console logged the `user` callback returned from the `AuthState` observable that is sent to the `query$`, and I can see in my browser that my user information is logged there. What could possibly be a reason why the `this.user$` subscription on my page is not triggered now?

I use the following to register my users when they click submit:

**Signup.page.ts**

```
async onSignUp() {
    await this.auth.userSignUp(this.signUpForm);
    this.auth.tokenChanged.subscribe(
      (uid) => {
        this.db.query$('auth/register', {
          userUID: uid,
          username: this.signUpForm.value.username,
          mod: 0
        }).subscribe((result) => {
            this.route.navigateByUrl('tabs/profile');
        })
      }
    )
  }
```

The auth methods method called above are found in my `AuthService`. `TokenChanged` is a `Subject` that can be subscribed to, which is how I know in the above method that my user is created and that I can save the user to my database.

```
  async userSignUp(signUpForm: FormGroup): Promise<string> {
    await this.afAuth.auth.createUserWithEmailAndPassword(signUpForm.value.email, signUpForm.value.password).then(
      () => {
        this.afAuth.auth.currentUser.sendEmailVerification().then(
          () => {
            this.token = this.afAuth.auth.currentUser.uid;
            this.tokenChanged.next(this.token);
          }
        );
      }
    );
    return this.token;
  }
```

This one has me stumped! What am I missing? Thank you!

**Comment by alberthoekstra** [comment]
try this:
```
  async userSignUp(signUpForm: FormGroup): Promise<string> {
    await this.afAuth.auth.createUserWithEmailAndPassword(signUpForm.value.email, signUpForm.value.password);
    await this.afAuth.auth.currentUser.sendEmailVerification();
    this.token = this.afAuth.auth.currentUser.uid;
    this.tokenChanged.next(this.token);
    return this.token;
  }
```

**Comment by uncvrd** [comment]
@alberthoekstra thanks for the suggestion! No luck there unfortunately. I did make some progress and maybe my concept of observables is off but I have the following observable on two different tabs: Home and Search.

```
this.auth.user$.subscribe((user) => {
    console.log(user)
})
```

When the app loads up, the home page is loaded and I subscribe to `this.auth.user$` in my `ngOnInit` like seen above. I route to my profile tab and register. I go back to my home page and nothing is logged to my screen, but when I visit my home page for the first time (which also has the `this.auth.user$` on init) it prints out my user info as expected. 

But what I don't understand is, I do the same exact thing with the login feature from AngularFire. I start on the Home Page -> go to profile page and log in -> and when I route to the home page for the first time, all my data is loaded (**this part doesn't happen when I use the register function**)

I can make a video of this if it is confusing haha

**Comment by uncvrd** [comment]
@alberthoekstra so I made more progress and would like your opinion on best practices with the `authState`. In my `AuthService` constructor I initially had the following, which queries for a firebase document:

```
this.user$ = this.afAuth.authState.pipe(
      switchMap(
        (user) => {
          return user ? db.doc$(`users/${user.uid}`) : of(null)
        } 
      )
    );
```

In my `DbService` you can see the method here:

```
  doc$(path): Observable<any> {
    return this.afs
      .doc(path)
      .snapshotChanges()
      .pipe(
        map(doc => {
          return { id: doc.payload.id, ...doc.payload.data() };
        })
      );
  }
```

And when I register, the `authState` changes which means that observable in my `AuthService` constructor is triggered again. **However, when I change to my custom api backend it stops working**

This is now my `authState` observable in my `AuthService:

```
this.user$ = this.afAuth.authState.pipe(
      switchMap(
        (user) => {
          return user ? db.query$(`user/find?userUID=${user.uid}`) : of(null)
        } 
      )
    );
```

Where `query$` is an HttpClient GET request to my db.

```
  query$(path: string, data?: Object) {
    if(data) {
      return this.http.post(`${this.apiUrl}/${path}`, data)
    } else {
      return this.http.get(`${this.apiUrl}/${path}`);
    }
  }
```

My theory is that when the `authState` is changed, the data in the database hasn't been added yet so when it's triggered to update the `this.user$` variable in `AuthService`, it still doesn't find anything to return. Unlike Firestore document subscriptions which pushes an update to the observable when something changes in the database. A normal `HttpClient` GET only sends one response to the observable on the initial GET request. Is this theory correct?

**Comment by gubbigubbi** [comment]
I too am having this issue, I wonder if it has something to do with Ionic's tabs. I also tried unsubscribing to the user observable but still no luck...

**Comment by gubbigubbi** [comment]
I figured it out! For some reason you need to add take(1), like so:
    this.user$ = afAuth.authState.pipe(
      take(1),
      switchMap(user => {
        return user
          ? this.fb.doc<User>(`users/${user.uid}`).valueChanges()
          : of(null);
      }),
    );

I believe the take(1) cause the observable to unsubscribe between tab changes?

**Comment by jimmykane** [comment]
This should be fixed with the latest version

**Comment by gubbigubbi** [comment]
Confirmed - this is fixed in 5.4.2
